### PR TITLE
add some types from cls_rgw_ops.h to ceph-dencoder

### PR DESCRIPTION
**Pull request by zhengp570**
New classes are declared in #include "cls/rgw/cls_rgw_ops.h"
Missing memeber functions are declared and defined
Fixes: https://tracker.ceph.com/issues/54054
Signed-off-by: zhengp570 <zhengp570@gmail.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
  - [ ] Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/cls/rgw/cls_rgw_ops.cc [hunk 1]
```diff
@@ -388,6 +388,82 @@ void rgw_cls_bucket_update_stats_op::dump(Formatter *f) const
...
+
+void rgw_cls_obj_store_pg_ver_op::dump(Formatter *f) const {
+  ::encode_json("attr", attr, f);
+}
+
+void rgw_cls_obj_store_pg_ver_op::generate_test_instances(list<rgw_cls_obj_store_pg_ver_op *>& o){
+  rgw_cls_obj_store_pg_ver_op *r = new rgw_cls_obj_store_pg_ver_op;
+  r->attr = "test";
+  o.push_back(r);
+
```
Comment (iqbalredkhan) [inline src/cls/rgw/cls_rgw_ops.cc:416]: Extra Line..

File: src/cls/rgw/cls_rgw_ops.cc [hunk 2]
```diff
@@ -388,6 +388,82 @@ void rgw_cls_bucket_update_stats_op::dump(Formatter *f) const
...
+  }
+  ::encode_json("keep_attr_prefix", k, f);
+}
+
+void rgw_cls_obj_remove_op::generate_test_instances(list<rgw_cls_obj_remove_op *>& o) {
+  rgw_cls_obj_remove_op *r = new rgw_cls_obj_remove_op;
+  r->keep_attr_prefixes.push_back("111");
+  r->keep_attr_prefixes.push_back("222");
+  o.push_back(r);
+
```
Comment (iqbalredkhan) [inline src/cls/rgw/cls_rgw_ops.cc:404]: Extra space

File: src/cls/rgw/cls_rgw_ops.cc [hunk 3]
```diff
@@ -388,6 +388,82 @@ void rgw_cls_bucket_update_stats_op::dump(Formatter *f) const
...
+  ::encode_json("check_prefix", check_prefix, f);
+  ::encode_json("fail_if_exist", fail_if_exist, f);
+}
+
+void rgw_cls_obj_check_attrs_prefix::generate_test_instances(list<rgw_cls_obj_check_attrs_prefix *>& o){
+  rgw_cls_obj_check_attrs_prefix *r = new rgw_cls_obj_check_attrs_prefix;
+  r->check_prefix = "test_prefix";
+  r->fail_if_exist = true;
+  o.push_back(r);
+
```
Comment (iqbalredkhan) [inline src/cls/rgw/cls_rgw_ops.cc:430]: Same here

File: src/cls/rgw/cls_rgw_ops.cc [hunk 4]
```diff
@@ -388,6 +388,82 @@ void rgw_cls_bucket_update_stats_op::dump(Formatter *f) const
...
+  o.push_back(new rgw_cls_obj_check_attrs_prefix);
+}
+
+void rgw_cls_obj_check_mtime::dump(Formatter *f) const {
+  ::encode_json("mtime", mtime, f);
+  ::encode_json("type", int(type), f);
+  ::encode_json("high_precision_time", high_precision_time, f);
+}
+
+void rgw_cls_obj_check_mtime::generate_test_instances(list<rgw_cls_obj_check_mtime *>& o) {
```
Comment (iqbalredkhan) [inline src/cls/rgw/cls_rgw_ops.cc:440]: Hey @cbodley , I have reviewed these changes, and all seems fine but I have difficulties in **void rgw_cls_obj_check_mtime::generate_test_instances(list<rgw_cls_obj_check_mtime *>& o)** function.
Please check once

File: src/cls/rgw/cls_rgw_ops.cc (hunk 1 above)
Comment (zhengp570) [reply src/cls/rgw/cls_rgw_ops.cc]: Thanks, I have removed the line.

**Pull request closed without merging**